File: LintCode/cpp/Gray Code.cc
class Solution {
public:
    /**
     * @param n a number
     * @return Gray code
     */
    vector<int> grayCode(int n) {
        if (n == 0) {
            return {0};
        }
        if (n == 1) {
            return {0, 1};
        }
        vector<int> res = {0, 1, 3, 2};
        
        for (int i = 3; i <= n; i++) {
            int num = (1 << i) / 2;
            for (int j = num - 1; j >= 0; j--) {
                res.push_back((1 << (i - 1)) | res[j]);
            }
        }
        return res;
    }
};

File: Leetcode/Algorithm/python/Gray Code.py
class Solution(object):
    def grayCode(self, n):
        if n == 0:
            return [0]
        gray_code = ['0', '1']
        for i in xrange(1, n):
            gray_code = self.get_gray_code(gray_code)
        
        return map(lambda x: int(x, 2), gray_code)
        
    def get_gray_code(self, gray_code):
        a = map(lambda x: '0' + x, gray_code)
        b = map(lambda x: '1' + x, gray_code[::-1])
        return a + b
        
